solo-io/gloo
# Fix non-persisted proxy validation in glooctl check

# Description
 - Updates `glooctl check` to use the GPRC endpoint for proxy validation, ala `glooctl get proxies`.
 - This fixes an issue where the check would report false-positives for invalid proxies that were not persisted.
   - Prior to this change, glooctl check used a k8s client to obtain proxy resources
   - Since 1.12.x, proxy resources are not persisted by default, so this k8s client failed to obtain proxies, and in turn failed to report errors on proxy resources

# Validating these changes
 - Build glooctl locally
    - `make build-cli-local`
    - for me, this is equivalent to `make glooctl-darwin-amd64`. This will differ depending on your OS and system architechture. I will use `glooctl-darwin-amd64` as a stand-in for the built glooctl binary for the rest of this PR description
  - Install Gloo and set up the failing resources
    - `helm repo update`
    - `helm install gloo gloo/gloo -n gloo-system --create-namespace --devel --version v1.15.0-beta6`
    - ```
      kubectl apply -f - <<EOF
      apiVersion: gloo.solo.io/v1
      kind: Upstream
      metadata:
        name: httpbin
        namespace: gloo-system
      spec:
        static:
          hosts:
            - addr: httpbin.org
              port: 80
      EOF
      ```
    - ```
      kubectl apply -f - <<EOF
      apiVersion: gateway.solo.io/v1
      kind: VirtualService
      metadata:
        name: test-static
        namespace: gloo-system
      spec:
        virtualHost:
          domains:
            - '*'
          routes:
            - matchers:
              - prefix: /inja
              options:
                prefixRewrite: /
                transformations:
                  responseTransformation:
                    transformationTemplate:
                      body:
                        text: '{% if int(header("content-length")) > 2000 %} {"error" : "Too
                          Damn Big!!"} {% else %} {{ body() }} {% endif %}'
                      headers:
                        :status:
                          text: '{% if int(header("content-length")) > 2000 %}401{% else %}{{header(":status")
                            }}{% endif %}'
                      parseBodyBehavior: DontParse
              routeAction:
                single:
                  upstream:
                    name: httpbin
                    namespace: gloo-system
      EOF
      ```
  - Update the virtual service so that the body transformation is no longer valid (e.g., remove the leading `{`)
  - Use the newly built binary and confirm that an error with the proxy is reported
    - `./_output/glooctl-darwin-amd64 check`
  - Use system glooctl install and confirm that an error with the proxy is not reported
    - `glooctl check`
  - set `gateway.persistProxySpec` to `true` in the system settings resource
  - Use the newly built binary and confirm that an error with the proxy is still reported
    - `./_output/glooctl-darwin-amd64 check`
  - Use the system glooctl install, and confirm that an error with the proxy is reported, and that the output matches that of the newly built binary exactly
    - `glooctl check`
